lib/src/perf_benchmark_base.dart

```diff
@@ -0,0 +1,146 @@
+    await _createFifos();
+    try {
+      perfProcess = await Process.start('perf', [
+        'stat',
+        '--delay',
+        '-1',
+        '--control',
+        'fifo:$perfControlFifo,$perfControlAck',
+        '-j',
+        '-p',
+        '$pid'
+      ]);
+      //await Future<void>.delayed(const Duration(seconds: 2));
+
+      openedFifo = File(perfControlFifo).openSync(mode: FileMode.writeOnly);
+
+      openedAck = File(perfControlAck).openSync();
+      openedFifo.writeStringSync('enable\n');
+      _waitForAck();
+    } catch (e) {
+      await fifoDir.delete(recursive: true);
+      rethrow;
+    }
+  }
+
+  Future<void> _stopPerfStat(int totalIterations) async {
+    try {
+      openedFifo.writeStringSync('disable\n');
+      openedFifo.closeSync();
+      _waitForAck();
+      openedAck.closeSync();
+      perfProcess.kill(ProcessSignal.sigint);
+      final lines =
+          utf8.decoder.bind(perfProcess.stderr).transform(const LineSplitter());
+      // Exit code from perf is -2 when terminated with SIGINT.
+      final exitCode = await perfProcess.exitCode;
+      if (exitCode != 0 && exitCode != -2) {
+        throw ProcessException(
+            'perf',
+            [
+              'stat',
+              '--delay',
+              '-1',
+              '--control',
+              'fifo:$perfControlFifo,$perfControlAck',
+              '-j',
+              '-p',
+              '$pid'
+            ],
+            (await lines.toList()).join('\n'),
+            exitCode);
+      }
+      final events = [
+        await for (final line in lines)
+          if (line.startsWith('{"counter-value" : '))
+            jsonDecode(line) as Map<String, dynamic>
+      ];
+      _reportPerfStats(events, totalIterations);
+    } finally {
+      await fifoDir.delete(recursive: true);
+    }
+  }
+
+  /// Measures the score for the benchmark and returns it.
+  Future<double> measurePerf() async {
+    setup();
+    // Warmup for at least 100ms. Discard result.
+    measureForImpl(warmup, 100);
+    await _startPerfStat();
+    // Run the benchmark for at least 2000ms.
+    var result = measureForImpl(exercise, minimumMeasureDurationMillis);
+    await _stopPerfStat(result.totalIterations);
+    teardown();
+    return result.score;
+  }
+
+  Future<void> reportPerf() async {
+    emitter.emit(name, await measurePerf());
+  }
+
+  void _waitForAck() {
+    var ack = <int>[...openedAck.readSync(5)];
+    while (ack.length < 5) {
+      ack.addAll(openedAck.readSync(5 - ack.length));
+    }
+    if (String.fromCharCodes(ack) != 'ack\n\x00') {
+      print('Ack was $ack');
+    }
+  }
+
+  void _reportPerfStats(List<Map<String, dynamic>> events, int iterations) {
+    for (final {'event': String event, 'counter-value': String counterString}
+        in events) {
+      final metric =
+          {'cycles:u': 'CpuCycles', 'page-faults:u': 'MajorPageFaults'}[event];
+      if (metric != null) {
+        emitter.emit(
```

Comment:
The built-in PrintEmitter is adding " us." after all metrics it prints out. This is incorrect for the perf metrics and the totalIterations metric. It is also printing the "RunTime" metric on all its lines, instead of the correct metric name.

Should the ScoreEmitter class be change to take an optional unit parameter and metric name? Ideas about how to refactor this most cleanly are welcome.

Reply:
ScoreEmitter.emit now has optional metric and unit arguments.